src/types/Environment.cpp

```diff
@@ -9,30 +9,42 @@ using namespace ::Microsoft::Console::Utils;
 // We cannot use spand or not_null because we're dealing with \0\0-terminated buffers of unknown length
 #pragma warning(disable : 26481 26429)
 
+// Function Description:
+// - Wraps win32's CreateEnvironmentBlock to return a smart pointer.
+EnvironmentBlockPtr Microsoft::Console::Utils::CreateEnvironmentBlock()
+{
+    void* newEnvironmentBlock{ nullptr };
+    if (!::CreateEnvironmentBlock(&newEnvironmentBlock, GetCurrentProcessToken(), FALSE))
+    {
+        return nullptr;
+    }
+    return EnvironmentBlockPtr{ newEnvironmentBlock };
+}
+
 // Function Description:
 // - Updates an EnvironmentVariableMapW with the current process's unicode
 //   environment variables ignoring ones already set in the provided map.
 // Arguments:
 // - map: The map to populate with the current processes's environment variables.
+// - environmentBlock: Optional environment block to use when filling map. If omitted,
+//   defaults to the current environment.
 // Return Value:
 // - S_OK if we succeeded, or an appropriate HRESULT for failing
-HRESULT Microsoft::Console::Utils::UpdateEnvironmentMapW(EnvironmentVariableMapW& map) noexcept

```

Comment:
It seems a bit strange to take `std::optional<void*>` and then treat `nullptr` as equivalent to `std::nullopt`. I don't see `std::optional<T*>` anywhere else in the source tree. But perhaps it makes sense if you consider `nullptr` an invalid argument here.

Reply:
I concur - why isn't the parameter just a `EnvironmentBlockPtr` that we compare to `nullptr` to check if it was provided?